plugins/overwatch.rb

```diff
@@ -0,0 +1,660 @@
+              process(sid)
+            end
+
+          # Error handling - Yea, lots of stuff in here to keep errors from popping up in the console when sessions die or go stale. 
+          # We also skip sessions below that have more than three errors because they slow down processing.
+          rescue ::Exception => e
+            if e.class.to_s == "Rex::TimeoutError"  # Error when a session stops responding and the requests time out.
+              session_log(sid, "Timed out.")
+            elsif e.class.to_s == "Rex::Post::Meterpreter::RequestError" # Error when a session does not have access to routing information (PHP/Meterpreter sessions throw this error.
+              if self.state[sid][:error_count]
+                self.state[sid][:error_count] = self.state[sid][:error_count] + 1
+              else
+                self.state[sid][:error_count] = 1
+              end
+              session_log(sid, "Access error: #{self.state[sid][:error_count].to_s} of 3. Possibly migrating or routing not available.") # Only log this error three times
+              session_log(sid, "No access to routing, skipping session.") if self.state[sid][:error_count] >= 3 # Skip this session after it errored three times
+            else
+              session_log(sid, "Triggered an exception: #{e.class} <> #{e} #{e.backtrace}") # Log any other errors.
+            end
+          end
+        end
+        sleep(0.5) # This sleep time seams to work the best to keep processes from stepping on each other.
+      end
+    end
+
+    # Manages the top level of processing the sessions.
+    # Sessions must be fully operational before the subprocesses can work.
+    #
+    # @sid [int class] Session ID of the current session being processed
+    #
+    # @return [void] A useful return value is not expected here
+    def process(sid)
+      return unless framework.sessions[sid].info #Make sure session is fully active before processing
+      self.state[sid] ||= {}
+      store_session_info(sid)
+      process_routing(sid)
+      process_sessions(sid)
+    end
+
+    # Initialization and update session information.
+    #
+    # @sid [int class] Session ID of the current session
+    #
+    # @return [void] A useful return value is not expected here
+    def store_session_info(sid)
+      self.state[sid][:last_update] = Time.now.to_f
+      return if self.state[sid][:initialized]
+      self.state[sid][:info] = framework.sessions[sid].inspect
+      session_log(sid, "registered")
+      self.state[sid][:initialized] = true
+      self.state[sid][:error_count] = 0
+    end
+
+    # Prints lines of the log along with session information.
+    # Only prints to log if logging is enabled.
+    #
+    # @sid [int class] Session ID of the current session
+    # @msg [string class] Message to be displayed in the log
+    #
+    # @return [void] A useful return value is not expected here
+    def session_log(sid, msg)
```

Comment:
Any reason not to use the native logging functionality? Makes it easier to deal with when we change things at the lower levels (which i'm planning to do, async stack traces can get messy today).